### PR TITLE
Unify `resolve` calls

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

This unifies the recursive calls for `resolve`. Previously, we had a bunch of free-standing functions scattered all over the place. This PR solves this by cleaning up those sections.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/0/1202912966917502/f) _(internal)_

## 🔍 What does this change?

- Move `Resolve` implementation to `resolve` module (e.g. `entity::resolve` contains `impl Resolve for EntityRecord`)
- Make free-standing functions a `Resolve` implementation instead
- Rename `postgres::resolve` module to `postgres::context` (it does not contain any non-context related things anymore)
- Return an error on out-of-bounds indexing rather than panicking

NB: This deduplicates code, however, the diff has more lines added than removed because of the import statements at the top.

## 🛡 What tests cover this?

This is an internal change only, the same tests apply here as before

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/resolve.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+where
+    C: Sync + ?Sized,
+{
+    async fn resolve(&self, path: &[PathSegment], context: &C) -> Result<Literal, ResolveError> {
+        match path {
+            [] => todo!("{}", UNIMPLEMENTED_LITERAL_OBJECT),
+            [head_path_segment, tail_path_segments @ ..] => {
+                let uri = BaseUri::new(head_path_segment.identifier.clone())
+                    .into_report()
+                    .change_context(ResolveError::StoreReadError)?;
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/resolve.rs:97]: Why is this a `StoreReadError`?

File: packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/link_type/resolve.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                                        .await
+                                })
+                                .try_collect()
+                                .await?,
+                        ))
+                    }
+                    link_type_uri => {
+                        let versioned_uri = VersionedUri::from_str(link_type_uri)
+                            .into_report()
+                            .change_context(ResolveError::StoreReadError)?;
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/link_type/resolve.rs:46]: Why is this a `StoreReadError`?

File: packages/graph/hash_graph/lib/graph/src/store/query.rs [hunk 3]
```diff
@@ -346,6 +347,13 @@ impl fmt::Display for ResolveError {
                 write!(fmt, "cannot index `{literal:?}` with path `{path}`")
             }
             Self::StoreReadError => fmt.write_str("could not read data from store"),
+            Self::OutOfBounds { index, list } => {
+                write!(
+                    fmt,
+                    "index out of bound, got index of `{index}`, but a length of `{}`",
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/store/query.rs:353]:
```suggestion
"index out of bounds, requested index was `{index}`, but only a length of `{}`",
```

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/resolve.rs (hunk 1 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/resolve.rs]: I "fixed" this in 6d03a3fe953e3a37bbb69766fd0619b3d9117f76, however, `ResolveError` is badly designed, instead it should be separated structs as base errors. This was an oversight when cleaning up after initially implementing the queries. You want me to change this in this PR or in a follow up?

Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/resolve.rs]: In the interest of progress 💪 lets do it in a follow-up 🙃

**Review by Alfred-Mountfield — APPROVED** [review]

**Pull request merged**